gno.land/pkg/sdk/vm/keeper.go

```diff
@@ -135,6 +143,92 @@ func (vm *VMKeeper) getGnoStore(ctx sdk.Context) gno.Store {
 	}
 }
 
+// Namespace can be either a user or crypto address.
+// XXX: Uppercase should be valid but transform into lowercase ?
+var reNamespace = regexp.MustCompile(`^gno.land/(?:r|p)/([a-zA-Z]+[_a-zA-Z0-9]+)`)
+
+// checkNamespacePermission check if the user as given has correct permssion to on the given pkg path
+func (vm *VMKeeper) checkNamespacePermission(ctx sdk.Context, creator crypto.Address, pkgPath string) error {
+	const sysUsersPkg = "gno.land/r/demo/users"
+
+	store := vm.getGnoStore(ctx)
+
+	// if `sysUsersPkg` does not exists -> skip validation.
+	usersPkg := store.GetPackage(sysUsersPkg, false)
+	if usersPkg == nil {
+		return nil
+	}
+
+	// XXX: is this necessary ?
+	pkgPath = filepath.Clean(pkgPath) // cleanup pkgpath
+
+	match := reNamespace.FindStringSubmatch(pkgPath)
+	if len(match) != 2 {
+		return fmt.Errorf("%w: %q", ErrInvalidRealm, pkgPath)
+	}
+	username := match[1]
+
+	// Lowercase username
+	username = strings.ToLower(username)
```

Comment:
Interesting. I'd like more opinions; I believe it's a good idea to enforce lowercase usernames.
This check makes sense to exist, but I think we should maybe enforce it from `r/users` too.

Reply:
```suggestion
	namespace = strings.ToLower(namespace)
```

Let's use `namespace` in the implementation. Today we use `r/users`, tomorrow, we'll add team support (a.k.a., organizations).

Reply:
Yes, I have some doubts about this. I understand that the username should be in lowercase to avoid confusion, but I feel that `gno.land/r/guiland/hello` and `gno.land/r/gUiLaDd/hello` should, in that case, be case-insensitive, at least for the username part.

Reply:
I believe that `r/users` will not only be used for namespaces but also for free-text mode, such as "Hey @gfanton" in `r/boards`. In the future, we may also want to provide additional services that come with character restrictions, like a DNS system based on usernames.

In other words, it makes sense to limit `r/users` to accept usernames that match the pattern `[a-z][a-z0-9_]*` or similar. In the `pkgpath`, we will keep the underscore, but in the DNS, we will replace the underscore with a dash.

Let's use lowercase for now. Additionally, we should check/fix the `r/users` issue at the source to prevent having `N->1` `r/users` for namespaces.